### PR TITLE
Integrate PHPickerViewController in Site Icon picker

### PR DESCRIPTION
**Pull request by kean**
Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/21190.

- Replace the WPMediaPicker flow with a context menu
- Replace the action sheet with the menu (not feature flagged because it's a small visual change)
- Introduce `MediaPickerMenu` API for creating the media picker actions. It has a convenient API that ensures the actions are consistent across different parts of the app. The picker uses `UIImageViewContoller` directly (instead of `WPMediaCapturePresenter` from WPMediaPicker) – moving forward with the WPMediaPicker removal.

Screenshot 2023-08-21 at 7 50 04 AM

## To test:

> [!IMPORTANT]
> - Enable the "Native Photo Picker" feature flag before testing

**Choose from Device (Happy Path)”**

- Tap the site icon and select “Choose form Device”
- Verify that no permission dialog is presented, but you can still access the photos (requires fresh install)
- Verify that the picker shows only images
- Select a photo and verify that it shows the Crop screen on top of the picker
- Crop the photo, tap “Use”, and verify that the icon is uploaded

**Choose from Device (Misc)”**

- Verify that if you select a photo and tap “Cancel” on the Crop screen or drag it to dismiss it, you are back to the picker screen where you can change the selection
- Verify that after changing the selection and going through the Crop flow, the icon is uploaded

**Camera**
- Verify that the first time you open the camera, it asks for permissions (requires fresh install)
- Verify that if you tap “Take Photo”, but the access is restricted, the app shows an alert with an explanation and an “Open Settings” button. Verify the the button opens the Settings app.
- Take a photo and verify that the crop screen opens and verify that the cropped photo gets uploaded as a site avatar

**QuickStart**
- Install the app (fresh install) and login
- Create a new site
- Click “Show me Around” when asked
- Enable Quick Start for New Site (either via debug settings or by creating a new site)
- Reach the “Choose a unique site icon” step 
- Verify that New Site tour continuing after cancelling the Site Icon menu or going with one of the options

Recordings: [menu-cancelled.mp4](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/e4be0eb3-2869-4112-afef-7a110f9b3350), [icon-updated.mp4](https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/c90cc2b4-e0c1-4de4-bdc4-4982669277ed).

**Removing Icon**

- Set the icon
- Remove the icon
- Verify that the “Remove Site Icon” button is no longer visible

**Misc**
- Verify that “Create with Emoji” action is not displayed when siteIconCreator is disabled
- Verify that WebP images can be used (required special handling)

**Options**

- let mediaPickerMenu = MediaPickerMenu(presentingViewController: self) modify options and verify video filter and single selection

**Flag Disabled**

- Verify that then “Native Photo Picker” is disabled, the UIMenu is still used (for simplicity). I kept only once implementation to make it easier to refactor.

**Bonus Points**

- Test all combinations of `MediaPickerMenu` options (`filter` and `isMultipleSelectionEnabled`) 
- Verify that `MediaLibraryPickerDataSource` and pickers get deallocated after selection or cancellation

## Regression Notes
1. Potential unintended areas of impact: Site Icon selection flow
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual tests
3. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Blog/Site Picker/SitePickerViewController+SiteIcon.swift [hunk 1]
```diff
@@ -3,73 +3,79 @@ import WordPressFlux
...
-        alert.addDefaultActionWithTitle(SiteIconAlertStrings.Actions.changeSiteIcon) { [weak self] _ in
-            NoticesDispatch.unlock()
-            self?.updateSiteIcon()
-        }
-
-        if blog.hasIcon {
-            alert.addDestructiveActionWithTitle(SiteIconAlertStrings.Actions.removeSiteIcon) { [weak self] _ in
-                NoticesDispatch.unlock()
-                self?.removeSiteIcon()
+    func didShowSiteIconMenu() {
```
Comment (kean) [inline WordPress/Classes/ViewRelated/Blog/Site Picker/SitePickerViewController+SiteIcon.swift:22]: There doesn't seem to be any good way to track `UIMenu` dismiss event, so I implemented this "pull-based" check for the site icon flow termination. Not ideal, but works.

File: WordPress/Classes/ViewRelated/Blog/Blog Details/Detail Header/SiteIconView.swift [hunk 2]
```diff
@@ -40,6 +40,14 @@ class SiteIconView: UIView {
 
     private var dropInteraction: UIDropInteraction?
 
+    /// Set the menu to be displayed when the button is tapped. The menu replaces
+    /// teh default on tap action.
+    func setMenu(_ menu: UIMenu, onMenuTriggerd: @escaping () -> Void) {
```
Comment (momo-ozawa) [inline WordPress/Classes/ViewRelated/Blog/Blog Details/Detail Header/SiteIconView.swift]: nit
```suggestion
    func setMenu(_ menu: UIMenu, onMenuTriggered: @escaping () -> Void) {
```

File: WordPress/Classes/ViewRelated/Media/MediaPickerMenu.swift [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+            case .images: picker.mediaTypes = [UTType.image.identifier]
+            case  .videos: picker.mediaTypes = [UTType.movie.identifier]
+            }
+        } else {
+            picker.mediaTypes = [UTType.image.identifier, UTType.movie.identifier]
+        }
+
+        let delegate = ImagePickerDelegate(delegate: delegate)
+        picker.delegate = delegate
+        objc_setAssociatedObject(picker, &MediaPickerMenu.strongDelegateKey, delegate, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
```
Comment (momo-ozawa) [inline WordPress/Classes/ViewRelated/Media/MediaPickerMenu.swift:174]: Q: are you doing this to ensure the picker / delegate gets deallocated along with the menu?

**Review by momo-ozawa — APPROVED** [review]
I tested all the cases, and it works as described! Left some comments

File: WordPress/Classes/ViewRelated/Media/MediaPickerMenu.swift (hunk 3 above)
Comment (kean) [reply WordPress/Classes/ViewRelated/Media/MediaPickerMenu.swift:174]: I need to retain `ImagePickerDelegate` that gets created by this method. It gets deallocated along with the image picker.

**Pull request merged**